src/pybind/mgr/k8sevents/module.py

```diff
@@ -0,0 +1,1072 @@
+                del self.events[event_name]
+
+    def process(self, log_object):
+        
+        log.debug("[DBG] k8sevents processing : {}".format(str(log_object)))
+
+        event_out = False
+        unique_name = True
+
+        if log_object.msg_type == 'audit':
+            # audit traffic : operator commands
+            if log_object.msg.endswith('finished'):
+                log.debug("[DBG] k8sevents received command finished msg")
+                event_out = True
+            else:
+                # NO OP - ignoring 'dispatch' log records
+                return
+        
+        elif log_object.msg_type == 'cluster':
+            # cluster messages : health checks
+            if log_object.event_name:
+                event_out = True
+
+        elif log_object.msg_type == 'config':
+            # configuration checker messages 
+            event_out = True
+            unique_name = False
+            
+        elif log_object.msg_type == 'heartbeat':
+            # hourly health message summary from Ceph
+            event_out = True
+            unique_name = False
+            log_object.msg = str(self.config_watcher)
+
+        else:
+            log.warning("[WRN] k8sevents received unknown msg_type - {}".format(log_object.msg_type))
+
+        if event_out:
+            # we don't cache non-unique events like heartbeats or config changes
+            if not unique_name or log_object.event_name not in self.events.keys():
+                event = KubernetesEvent(log_entry=log_object,
+                                        unique_name=unique_name)
+                event.write()
+                log.debug("[DBG] event(unique={}) creation ended : {}".format(unique_name, event.api_status))
+                if event.api_success and unique_name:
+                    self.events[log_object.event_name] = event
+            else:
+                event = self.events[log_object.event_name]
+                event.update(log_object)
+                log.debug("[DBG] event update ended : {}".format(event.api_status))
+
+            self.prune_events()
+
+        else:
+            log.debug("[DBG] k8sevents ignored message : {}".format(log_object.msg))
+
+    def run(self):
+        log.info("[INF] Ceph event processing thread started")
+
+        while True:
+
+            try:
+                log_object = event_queue.get(block=False)
+            except queue.Empty:
+                pass
+            else:
+                try:
+                    self.process(log_object)
+                except Exception:
+                    self.health = "{} Exception at {}".format(
+                        sys.exc_info()[0].__name__,
+                        datetime.strftime(datetime.now(),"%Y/%m/%d %H:%M:%S")
+                    )
+                    log.exception(self.health)
+                    break
+            
+            if not self.can_run:
+                break
+
+            time.sleep(0.5)
```

Comment:
when no events are firing this is a very lightweight loop, right?

Reply:
yep. If there is nothing in the queue the loop is basically just hitting the sleep which prevents the loop from just spinning on a core.